server/channels/app/report.go

```diff
@@ -4,11 +4,136 @@
 package app
 
 import (
+	"bytes"
+	"encoding/csv"
+	"fmt"
 	"net/http"
 
 	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/i18n"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/platform/shared/filestore"
 )
 
+func (a *App) SaveReportChunk(format string, prefix string, count int, reportData []model.ReportableObject) *model.AppError {
+	switch format {
+	case "csv":
+		return a.saveCSVChunk(prefix, count, reportData)
+	}
```

Comment:
It seems weird to have support for `format` everywhere (with only one type), but then hard-code the `ReportableObject` with a `ToReport()` that returns a `[]string` that's probably (?) only useful for CSV. It feels like we're pre-designing for a future alternative format but we'll have to change a ton to support that anyway. Would we consider simplifying and removing the `format` handling altogether?

Reply:
I'll address the overarching theme here then.

I think regardless of what the report format is, as we discussed before we wanted to design a system that's going to allow for a resumable batch job flow. To me that will always, in some capacity follow this workflow:
1) Start a batch job
2) Load some piece of the data for the report
3) Generate a piece of the report, save the file
4) Check if there is more data, if so go to step 2
5) If not, proceed to compile all of the pieces
6) Send the completed piece to the user

Now of course, I can't argue that other formats may have some other small nuances that may not EXACTLY fit this model, but based on what I've understood about reporting in the past, most of them will. PDFs for example, would likely be generated page-by-page, and then sewn together at the end. Same would likely go for DOCX, perhaps LaTeX as well. XLS is similar to CSV, so it would also follow the same flow.

I could continue with other formats, but I think it's important to make an effort to define a "line of best fit" here based on the information we have. I would like further reporting work to follow this flow, until there is a compelling reason to change it. And there is a possibility that will happen, but this to me feels like the best way to start.

As for the specific case above, we will always want to translate the data we have into a human readable form of some kind, which would likely be some arbitrary string, formatted as we see fit. I've added the hook per format as we may want to change that based on how we export. Perhaps the best use case for PDF would be a proper human-readable date string, whereas CSV or XLS would be best stored as something read best by MS Excel or other spreadsheet applications.

tl;dr I'm 4/5 on keeping the current formatting, as I believe this is likely the best way to handle report generation for almost all cases.

Reply:
Do you have an estimate for how soon we expect to start supporting other formats? CSV being what it is, I'm wondering if shipping CSV now and for the forseeable future would more than cover the required customer scenarios.

I'm only 1/5 on this given the code exists for now, but I'm always hesitant seeing preemptive complexity when N=1.

Reply:
So phase 1 of this work was only to support CSV, but the original proposal was to include at least PDF reporting. I don't believe following phases have been scheduled yet, but I assume they will be coming down the pipeline, and possibly sooner once the feature goes live, as I can see customers wanting something easier to visualize for non-technical users that comes out "ready-to-go".

Additionally, I'll reiterate that I think it's important to set up a system for how we'd want this to work going forward when we support other formats as I would assume it will be "when", not "if" we do. So I'd like to keep this design as it is to be prescriptive about what it means to add another report format (ie. add the case statements, add the hooks, create a job and go)

Reply:
> when we support other formats as I would assume it will be "when", not "if" we do

No objection to keeping thing as is, but my assumption would be "if" not "when". Plus, if we could export the raw data, we could easily build additional tools to turn that into different kinds of reports vs. having to extend the core product all the time.

Reply:
I think that might end up being a valid case, but I know the core reason we started this project was to provide a lot of those tools as a part of the core product since customers were already doing a lot of the reporting themselves, which is why I made my assumption. Direction may change, but I guess we'll see what happens.